cmd/relay-s3/main.go

```diff
@@ -0,0 +1,194 @@
+var (
+	version   = "master"
+	gitCommit = "none"
+	buildDate = "unknown"
+	log       = dlog.DefaultLogger
+)
+
+func main() {
+	app := &cli.App{
+		Name:     "drand-relay-s3",
+		Version:  version,
+		Usage:    "AWS S3 relay for randomness beacon",
+		Commands: []*cli.Command{runCmd},
+	}
+	cli.VersionPrinter = func(c *cli.Context) {
+		fmt.Printf("drand AWS S3 relay %v (date %v, commit %v)\n", version, buildDate, gitCommit)
+	}
+
+	err := app.Run(os.Args)
+	if err != nil {
+		fmt.Printf("error: %+v\n", err)
+		os.Exit(1)
+	}
+}
+
+var runCmd = &cli.Command{
+	Name:  "run",
+	Usage: "start a drand AWS S3 relay process",
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:    "chain-hash",
+			Usage:   "hex encoded hash of the drand group chain (optional)",
+			Aliases: []string{"chain"},
+		},
+		&cli.StringFlag{
+			Name:     "bucket",
+			Usage:    "name of the AWS bucket to upload to",
+			Aliases:  []string{"b"},
+			Required: true,
+		},
+		&cli.StringFlag{
+			Name:    "region",
+			Usage:   "name of the AWS region to use (optional)",
+			Aliases: []string{"r"},
+		},
+		&cli.StringFlag{
+			Name:    "grpc-connect",
+			Usage:   "host:port to dial to a drand gRPC API",
+			Aliases: []string{"connect", "c"},
+		},
+		&cli.StringSliceFlag{
+			Name:    "http-connect",
+			Usage:   "URL(s) of drand HTTP API(s) to relay",
+			Aliases: []string{"u"},
+		},
+		&cli.StringFlag{
+			Name:  "cert",
+			Usage: "file containing GRPC transport credentials of peer (optional)",
+		},
+		&cli.BoolFlag{
+			Name:  "insecure",
+			Usage: "allow insecure GRPC connection (optional)",
+		},
+		&cli.StringFlag{
+			Name:  "metrics",
+			Usage: "local host:port to bind a metrics servlet (optional)",
+		},
+	},
+
+	Action: func(cctx *cli.Context) error {
+		if cctx.IsSet("metrics") {
+			metricsListener := metrics.Start(cctx.String("metrics"), pprof.WithProfile(), nil)
+			defer metricsListener.Close()
+		}
+
+		var err error
+		var chainHash []byte
+
+		if cctx.String("chain-hash") != "" {
+			chainHash, err = hex.DecodeString(cctx.String("chain-hash"))
+			if err != nil {
+				return fmt.Errorf("decoding chain hash: %w", err)
+			}
+		}
+
+		sess, err := session.NewSession(&aws.Config{Region: aws.String(cctx.String("region"))})
+		if err != nil {
+			return fmt.Errorf("creating aws session: %w", err)
+		}
+
+		_, err = sess.Config.Credentials.Get()
+		if err != nil {
+			return fmt.Errorf("checking credentials: %w", err)
+		}
+
+		uploader := s3manager.NewUploader(sess)
+
+		var w client.Watcher
+
+		if len(cctx.StringSlice("http-connect")) > 0 {
+			w, err = newHTTPWatcher(cctx.StringSlice("http-connect"), chainHash)
+		} else if cctx.String("grpc-connect") != "" {
+			w, err = newGRPCWatcher(cctx.String("grpc-connect"), cctx.String("cert"), cctx.Bool("insecure"))
+		} else {
+			return errors.New("missing GRPC or HTTP address(es)")
+		}
+		if err != nil {
+			return err
+		}
+
+		ctx := context.Background()
+		for res := range w.Watch(ctx) {
```

Comment:
maybe put this in a for loop where you'll re-watch if an error back to the server does occur (as it can sometimes right now as i understand it). The only reason you want to stop is the `ctx` closing, not that Watch has closed to signal it missed a round

Reply:
hmm, I think we just log an error and keep going, I'll double check